Add oidc token support

salrashid123 opened the pull request:
PR adds oidc token support for this provider.   
see [https://github.com/terraform-providers/terraform-provider-google/issues/5669](https://github.com/terraform-providers/terraform-provider-google/issues/5669)

This pr is now complete and does has working testcases;

salrashid123 commented:
I updated the `go.mod` but still see conflict for some reason.

Anyway, i've reverified its functionality.  if the testcases are sufficient (which i actually just copied from the existing `data_source_google_service_account_access_token_test`, can we proceed w/ review?

danawillow commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,78 @@
+  target_audience = "%s"
+}
+
+output "oidc-token" {
```
You shouldn't need to have the output in the test, you can just verify by checking the data source directly

salrashid123 replied:
ok, i removed the explicit 'output` step, PTAL

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+This data source provides a google OpenID Connect (`oidc`) `id_token` 
+
+For more information see
+[OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) as well as [Authenticating using Google OpenID Connect Tokens](https://github.com/salrashid123/google_id_token)
```
Is there not an official GCP docs site that explains this?

salrashid123 replied:
unfortunately, no, thats the only doc about google OIDC tokens...

we do have:
  https://developers.google.com/identity/protocols/oauth2/openid-connect
but thats just focued on oidc tokens for user flow (whcih isn't used here as such)

as an internal ref: see b/136495350

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
````diff
@@ -0,0 +1,128 @@
+  In this mode, a the datasource will the service account provided by `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_KEYFILE_JSON` environment variable. 
+  
+  Sample configuration:
+  ```yaml
````
```suggestion
  ```hcl
```
(here and elsewhere)

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
````diff
@@ -0,0 +1,128 @@
+  
+  Sample configuration:
+  ```yaml
+  provider google {}
````
We typically leave this out of examples unless the example has a specific reason for including it

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+There are several ways to use this data source to get an `id_token`:
+
+* Using ServiceAccount JSON credential file.
+  In this mode, a the datasource will the service account provided by `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_KEYFILE_JSON` environment variable. 
```
```suggestion
  In this mode, the datasource will use the service account provided by `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_KEYFILE_JSON` environment variable. 
```
I think that's what this sentence was trying to say?

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+
+There are several ways to use this data source to get an `id_token`:
+
+* Using ServiceAccount JSON credential file.
```
Typically we ## each example individually rather than bullet points: see https://github.com/terraform-providers/terraform-provider-google/blob/master/website/docs/r/compute_address.html.markdown for an example docs  page with multiple examples

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
````diff
@@ -0,0 +1,128 @@
+  }
+  ```
+
+* Using `Application Default Credentials` on ComputeEngine or Kubernetes Engine
````
```suggestion
* Using `Application Default Credentials` on Compute Engine or Kubernetes Engine
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
````diff
@@ -0,0 +1,128 @@
+  ```
+
+* Using `Application Default Credentials` on ComputeEngine or Kubernetes Engine
+  In this mode, the datastource will use the the [instance identity document](https://cloud.google.com/compute/docs/instances/verifying-instance-identity) to provide the `id_token.  The sample usage is the same as the mode above except no environment variables are needed.
````
```suggestion
  In this mode, the data source will use the the [instance identity document](https://cloud.google.com/compute/docs/instances/verifying-instance-identity) to provide the `id_token`.  The sample usage is the same as the mode above except no environment variables are needed.
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+  In this mode, the datastource will use the the [instance identity document](https://cloud.google.com/compute/docs/instances/verifying-instance-identity) to provide the `id_token.  The sample usage is the same as the mode above except no environment variables are needed.
+
+* Using Service Account Impersonation.
+  In this mode, the service account that runs terraform will attempt to impersonate another service account and then acquire the second service accounts `id_token`.  This mode is similar to `access_token` that is provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  Utilizing this mechanism is requ ...
```
```suggestion
  In this mode, the service account that runs terraform will attempt to impersonate another service account and then acquire the second service account's `id_token`.  This mode is similar to `access_token` that is provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  Utilizing this mechanism requires the IAM `TokenCreator` role to be granted to the the origin service account _on_ the target account.
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+* Using Service Account Impersonation.
+  In this mode, the service account that runs terraform will attempt to impersonate another service account and then acquire the second service accounts `id_token`.  This mode is similar to `access_token` that is provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  Utilizing this mechanism is requ ...
+
+  There are further two variations on how this mode is used depending on how the source token is acquired.
```
```suggestion
  There are two further variations on how this mode is used depending on how the source token is acquired:
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+
+  There are further two variations on how this mode is used depending on how the source token is acquired.
+
+  1. If the source token is provided directly as a variable using the requestors token, then grant the requestor the `TokenCreator` role on the target service account.  In the following case, the identity in context with `gcloud` should have the TokenCreator role on `impersonated-account@project.iam.gserviceaccount.com`
```
```suggestion
  1. If the source token is provided directly as a variable using the requestor's token, then grant the requestor the `TokenCreator` role on the target service account.  In the following case, the identity in context with `gcloud` should have the TokenCreator role on `impersonated-account@project.iam.gserviceaccount.com`
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+  1. If the source token is provided directly as a variable using the requestors token, then grant the requestor the `TokenCreator` role on the target service account.  In the following case, the identity in context with `gcloud` should have the TokenCreator role on `impersonated-account@project.iam.gserviceaccount.com`
+
+
+  Then run the terraform apply and specify the token directly
```
```suggestion
  Then run `terraform apply` and specify the token directly
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+
+The following arguments are supported:
+
+* `target_audience` (Required) - The audience claim value to to issued the `id_token` for.
```
I'm not sure what this sentence means

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+The following arguments are supported:
+
+* `target_audience` (Required) - The audience claim value to to issued the `id_token` for.
+* `access_token` (Optional) - Raw access_token to use as the source credential.  Specifying this value will use IAMCredentials API even if any other variable is set.
```
```suggestion
* `access_token` (Optional) - Raw access_token to use as the source credential.  Specifying this value will use the IAMCredentials API even if any other variable is set.
```

If this  is the case, do you want to make it impossible for people to set this along with other fields? If so, you can set `ConflictsWith` in the schema

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+
+* `target_audience` (Required) - The audience claim value to to issued the `id_token` for.
+* `access_token` (Optional) - Raw access_token to use as the source credential.  Specifying this value will use IAMCredentials API even if any other variable is set.
+* `target_service_account` (Optional) - The email of the service account bing impersonated. Used only when using impersonation mode.
```
```suggestion
* `target_service_account` (Optional) - The email of the service account being impersonated. Used only when using impersonation mode.
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+* `target_audience` (Required) - The audience claim value to to issued the `id_token` for.
+* `access_token` (Optional) - Raw access_token to use as the source credential.  Specifying this value will use IAMCredentials API even if any other variable is set.
+* `target_service_account` (Optional) - The email of the service account bing impersonated. Used only when using impersonation mode.
+* `delegates` (Optional) - Delegate chain of approvals needed to perform full impersonation. Specify the fully qualified service account name.   Used only when using impersonation mode.
```
```suggestion
* `delegates` (Optional) - Delegate chain of approvals needed to perform full impersonation. Specify the fully qualified service account name. Used only when using impersonation mode.
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,128 @@
+For more information see
+[OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) as well as [Authenticating using Google OpenID Connect Tokens](https://github.com/salrashid123/google_id_token)
+
+## Example Usage
```
It's probably worth thinking about how to be a bit more concise with the examples and descriptions of them. This is fairly wordy right now.

salrashid123 commented on google/data_source_google_service_account_id_token.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+func getCredentials(c *Config, clientScopes []string) (google.Credentials, error) {
```
moved the code into `config.go`.
I needed to make the function exportable though `GetCredentials()`

salrashid123 commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,78 @@
+		Steps: []resource.TestStep{
+			{
+				Config:  testAccCheckGoogleServiceAccountIdToken_datasource(targetAudience),
+				Destroy: true,
```
you mean the `Destroy: true` bit?  i'm not sure why i have it there, i just removed it

salrashid123 commented:
ok, i made a number of changes:

* removed the static `access_token` thing;   ther'es no way that woudl've worked since the library terraform uses underlying credentials into which you can add in a token like that.  I suppose if a static access_token is set directly, i could somehow  use `config.AccessToken` but i'll have to revert that later in code. It'd be confusing and a very error prone.  

* re:  conflicts.  it looks like the primary go.mod conflict is with `golang.org/x/net v0.0.0-20200324143707-d3edc9973b7e`  which i think comes through with `"google.golang.org/api/idtoken"`

* re: docs.  I refactored the whole doc set and removed a lot of the stuff.  I marked the comments you had as resolved since those sections were deleted. PTAL

* re: upstream/magic modules.  Lets finish the review here...i'd likely mess up the PR if i did complicated stuff like that!

danawillow commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,73 @@
+	})
+}
+
+func testAccCheckGoogleServiceAccountIdToken_datasource(targetAudience string) string {
```
I think it's also worth having a test that sets more of the available fields in the data source. It also looks from the way the docs and resource are structured that there's two distinct use cases for this data source. Can you make sure both use cases are tested?

salrashid123 replied:
addin basic impersonation and basic/regular usage testcases.

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+
+# google\_service\_account\id\_token
+
+This data source provides a google OpenID Connect (`oidc`) `id_token`.  Tokens issued using this datasource are typically used to call external services that accept OIDC tokens for authentication (e.g. [Google Cloud Run](https://cloud.google.com/run/docs/authenticating/service-to-service)). 
```
```suggestion
This data source provides a Google OpenID Connect (`oidc`) `id_token`.  Tokens issued from this data source are typically used to call external services that accept OIDC tokens for authentication (e.g. [Google Cloud Run](https://cloud.google.com/run/docs/authenticating/service-to-service)). 
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+[OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html#IDToken).
+
+## Example Usage - ServiceAccount JSON credential file.
+  Datasource will the service account provided by `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_KEYFILE_JSON` environment variable. 
```
```suggestion
  `google_service_account_id_token` will use the configured [provider credentials](https://www.terraform.io/docs/providers/google/guides/provider_reference.html#credentials-1)
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
````diff
@@ -0,0 +1,105 @@
+  }
+  ```
+
+## Example Usage - `Application Default Credentials`
````
This is really the same as the previous example in the end- I made a suggestion for how to merge the two.

salrashid123 replied:
must've missed the suggest; i guess you meant delete the dup so thats in here now

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+  Datasource will automatically use credentials provided by [Application Default Credentials](https://cloud.google.com/docs/authentication/production)
+
+## Example Usage - Service Account Impersonation.
+  Datasource will use background impersonated credentials provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  
```
```suggestion
  `google_service_account_id_token` will use background impersonated credentials provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+## Example Usage - Service Account Impersonation.
+  Datasource will use background impersonated credentials provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  
+
+  Note: to use the following, you must grant `target_service_account` the
```
Does it make sense to have this as an iam_member resource in the example?

salrashid123 replied:
i think that part is included in the docs about impersonation datasource though. ...it'll make it a bit long here to include the full flow.  IMO, the impersonation thing isn't that common vs plain svc account id token stuff

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+}
+
+data "http" cloudrun {
+  url = "https://your.cloud.run.app/"
```
Is there an existing cloud run resource that gets you this url?

salrashid123 replied:
you mean like as part of the test?   
https://www.terraform.io/docs/providers/google/r/cloud_run_service.html  does create the cloud run resource here.  (eg, you use that to deploy some small webapp with authentication enabled (whcih is default i think), then use the service account the test runs as to setup iam binding (then wait maybe 60s  or so,, then use this provider to invoke that app

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
````diff
@@ -0,0 +1,105 @@
+
+```hcl
+
+data "google_service_account_id_token" oidc {
````
```suggestion
data "google_service_account_id_token" "oidc" {
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+  target_audience = "https://your.cloud.run.app/"
+}
+
+data "http" cloudrun {
```
```suggestion
data "http" "cloudrun" {
```

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+  }
+}
+
+output "oidc_token" {
```
Since this example already shows the token being used, how valuable is it to also show it as an output?

salrashid123 replied:
i can remove it but its shown there so a user can "inspect" the token too (eg, plug it into a jwt decoder to see its claims, etc).  removed here now though

danawillow commented on website/docs/d/datasource_google_service_account_id_token.html.markdown:
```diff
@@ -0,0 +1,105 @@
+The following arguments are supported:
+
+* `target_audience` (Required) - The audience claim for the `id_token`.
+* `target_service_account` (Optional) - THe email of the service account being impersonated.  Used only when using impersonation mode.
```
```suggestion
* `target_service_account` (Optional) - The email of the service account being impersonated.  Used only when using impersonation mode.
```

salrashid123 commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,73 @@
+			return fmt.Errorf("Not found: %s", name)
+		}
+
+		// TODO: validate the token belongs to the service account
```
i'll do that later;   i first need to understand the the basic test cases work and then figure out the mocks needed here

salrashid123 commented on google/data_source_google_service_account_id_token.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	// If the source token is just an access_token, all we can do is use the iamcredentials api to get an id_token
+	if fmt.Sprintf("%s", reflect.TypeOf(ts)) == "oauth2.staticTokenSource" {
```
i'm not sure here.   The reflect bit below returns an actual function type where what i'd like is the underlying type whch isn't visible.  I'm open to ideas on how to improve this though
```golang
 reflect.TypeOf(oauth2.StaticTokenSource)
    --> func(*oauth2.Token) oauth2.TokenSource
``` 
https://github.com/golang/oauth2/blob/master/oauth2.go#L312-L326

salrashid123 commented on google/data_source_google_service_account_id_token.go:
```diff
@@ -0,0 +1,130 @@
+	if err != nil {
+		return fmt.Errorf("unable to retrieve TokenSource: %v", err)
+	}
+	tok, err = ts.Token()
```
yeah, that was pretty confusing...i reused one variable for two different things.  In the version just committed, i setup a new variable to initialize and hold the idtoken

salrashid123 commented on google/data_source_google_service_account_id_token.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+func dataSourceGoogleServiceAccountIdTokenRead(d *schema.ResourceData, meta interface{}) error {
+	var idToken string
```
i actually do need this variable but its all the way at the end.  i incorrectly reused the `tok` variable for two different things.  in the version now, i inspect the inbound/provided token and then use that to reissue an `idToken`

salrashid123 commented:
Not sure why the build failed on that documentation page (its not included in this PR)...maybe rerun travis again?

tbpg commented on google/config.go:
```diff
@@ -694,37 +694,54 @@ func (c *Config) synchronousTimeout() time.Duration {
+	}
+	return creds.TokenSource, nil
+}
+
+func (c *Config) GetCredentials(clientScopes []string) (googleoauth.Credentials, error) {
```
```suggestion
// staticTokenSource is used to be able to identify static token sources without reflection.
type staticTokenSource {
	oauth2.TokenSource
}

func (c *Config) GetCredentials(clientScopes []string) (googleoauth.Credentials, error) {
```

tbpg commented on google/config.go:
```diff
@@ -694,37 +694,54 @@ func (c *Config) synchronousTimeout() time.Duration {
 		token := &oauth2.Token{AccessToken: contents}
-		return oauth2.StaticTokenSource(token), nil
+
+		return googleoauth.Credentials{
+			TokenSource: oauth2.StaticTokenSource(token),
+		}, nil
```
```suggestion
		return googleoauth.Credentials{
			TokenSource: staticTokenSource(oauth2.StaticTokenSource(token)),
		}, nil
```

tbpg commented on google/data_source_google_service_account_id_token.go:
```diff
@@ -0,0 +1,128 @@
+
+	// If the source token is just an access_token, all we can do is use the iamcredentials api to get an id_token
+
+	if reflect.TypeOf(ts).String() == "oauth2.staticTokenSource" {
```
```suggestion
	if _, ok := ts.(staticTokenSource); ok {
```

salrashid123 commented:
@tbpg  Thanks for the suggestions!

@danawillow i'm not sure if the last `website/google.erb` change that appeared would help with the travis error or not.  should i rebase and retry?

salrashid123 commented:
yeah, i don't understand the docs error there...the `make website-test` exits correctly for me

```bash
$ make website-test

Found no broken links.

FINISHED --2020-06-29 14:41:45--
Total wall clock time: 12s
Downloaded: 323 files, 30M in 0.06s (515 MB/s)
tf-website-google-temp
make[1]: Leaving directory '/path/to/src/github.com/hashicorp/terraform-website'

$ echo $?
0
```

danawillow commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,102 @@
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckGoogleServiceAccountIdToken_datasource(targetAudience),
```
```suggestion
				Config: testAccCheckGoogleServiceAccountIdToken_basic(targetAudience),
```

danawillow commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,102 @@
+	})
+}
+
+func testAccCheckGoogleServiceAccountIdToken_datasource(targetAudience string) string {
```
```suggestion
func testAccCheckGoogleServiceAccountIdToken_basic(targetAudience string) string {
```

danawillow commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,102 @@
+	}
+}
+
+func TestAccDataSourceGoogleServiceAccountIdToken_basic(t *testing.T) {
```
This test fails:
```
------- Stdout: -------
=== RUN   TestAccDataSourceGoogleServiceAccountIdToken_basic
=== PAUSE TestAccDataSourceGoogleServiceAccountIdToken_basic
=== CONT  TestAccDataSourceGoogleServiceAccountIdToken_basic
TestAccDataSourceGoogleServiceAccountIdToken_basic: testing.go:674: Step 0 error: Check failed: Check 2/2 error: Not found: google_service_account_id_token.default
--- FAIL: TestAccDataSourceGoogleServiceAccountIdToken_basic (0.72s)
FAIL
```

danawillow commented on google/data_source_google_service_account_id_token_test.go:
```diff
@@ -0,0 +1,102 @@
+func testAccCheckServiceAccountIdTokenValue(name, value string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		ms := s.RootModule()
+		rs, ok := ms.Outputs[name]
```
The first test fails on this  line:
```
------- Stdout: -------
=== RUN   TestAccDataSourceGoogleServiceAccountIdToken_basic
=== PAUSE TestAccDataSourceGoogleServiceAccountIdToken_basic
=== CONT  TestAccDataSourceGoogleServiceAccountIdToken_basic
TestAccDataSourceGoogleServiceAccountIdToken_basic: testing.go:674: Step 0 error: Check failed: Check 2/2 error: Not found: google_service_account_id_token.default
--- FAIL: TestAccDataSourceGoogleServiceAccountIdToken_basic (0.72s)
FAIL
```

I think you want to use `ms.Resources` here and then pass it in `resourceName` in the test.

danawillow commented:
Oh, the website test failure is unrelated. Don't worry about it.

salrashid123 commented:
ok, i added in actual tests..but here's the catch:

the basic tests seem to work fine

```
export  GOOGLE_REGION=us-central1
export  GOOGLE_CLOUD_KEYFILE_JSON=/path/to/svc_account.json
export  GOOGLE_ZONE=us-central1-a
export  GOOGLE_PROJECT=someproject
export  GOOGLE_SERVICE_ACCOUNT=svcaccount1@someproject.iam.gserviceaccount.com


$ make testacc TEST=./google TESTARGS='-run=TestAccDataSourceGoogleServiceAccountIdToken_basic'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
go generate  ./...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccDataSourceGoogleServiceAccountIdToken_basic -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceGoogleServiceAccountIdToken_basic
=== PAUSE TestAccDataSourceGoogleServiceAccountIdToken_basic
=== CONT  TestAccDataSourceGoogleServiceAccountIdToken_basic
--- PASS: TestAccDataSourceGoogleServiceAccountIdToken_basic (2.34s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	2.521s
````

---

but the one with impersonation needs a delay due to iam propagation delay.  Any suggestion on how to do that ...i suppose a `null resource+sleep ~60s` somehow?

at the moment, the first run will do the IAM assocation but fail since the permissions to impersonate hasn't propagated

```
$ make testacc TEST=./google TESTARGS='-run=TestAccDataSourceGoogleServiceAccountIdToken_impersonation'
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccDataSourceGoogleServiceAccountIdToken_impersonation -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceGoogleServiceAccountIdToken_impersonation
=== PAUSE TestAccDataSourceGoogleServiceAccountIdToken_impersonation
=== CONT  TestAccDataSourceGoogleServiceAccountIdToken_impersonation
2020/06/29 23:40:19 [INFO] Authenticating using configured Google JSON 'credentials'...
    TestAccDataSourceGoogleServiceAccountIdToken_impersonation: testing.go:674: Step 0 error: errors during refresh:
        
        Error: googleapi: Error 403: The caller does not have permission, forbidden
--- FAIL: TestAccDataSourceGoogleServiceAccountIdToken_impersonation (2.42s)

```
which will fail but implictly apply an iam binding thats needed

```
$ gcloud iam service-accounts get-iam-policy  tf-bootstrap-service-account@project.iam.gserviceaccount.com
bindings:
- members:
  - serviceAccount:svcaccount1@project.iam.gserviceaccount.com
  - serviceAccount:tf-bootstrap-service-account@project.iam.gserviceaccount.com
  role: roles/iam.serviceAccountTokenCreator
```

but if you wait ~60s and rerun the tests, it works

```
$ make testacc TEST=./google TESTARGS='-run=TestAccDataSourceGoogleServiceAccountIdToken_impersonation'
--- PASS: TestAccDataSourceGoogleServiceAccountIdToken_impersonation (8.09s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	8.274s

```

salrashid123 commented:
@danawillow think its ready now;
i'm not sure how the acc tests are setup internally (i.,e if you use the same set of svc accounts on the same project), but in the event that these iam bindings are setup one and persist test to test, then the acc testsin this case would pass the impersonation one

danawillow commented:
I don't totally understand your most recent question, but any environment variable will be the same for all tests, and we often run tests in parallel. Does that answer the question?

For the question around waiting, take a look at this test, which I think had the same problem: https://github.com/terraform-providers/terraform-provider-google/blob/master/google/provider_test.go#L477

salrashid123 commented:
thx,  i'm what i meant by that is when i ran just the impersonation acc test locally, it would fail the first time but associate the IAM binding w/ the service accounts i specified in the env-var.  If i ran the same test again, the test would pass since the binding was done a couple mins earlier.

I believe this step does that per test run:

https://github.com/terraform-providers/terraform-provider-google/blob/master/google/bootstrap_utils_test.go#L164

that bit is what allows the following _existing_ test to work
* https://github.com/terraform-providers/terraform-provider-google/blob/master/google/data_source_google_service_account_access_token_test.go#L52

```make testacc TEST=./google TESTARGS='-run=TestAccDataSourceGoogleServiceAccountAccessToken_basic'```


i guess what i'm trying to say is if the acc tests you're running now allows the test above to run, the impersonation test i've got in this PR should also work (i.,e they both require the same iam settings).  Could you rerun the acc tests?

```
make testacc TEST=./google TESTARGS='-run=TestAccDataSourceGoogleServiceAccountIdToken_impersonation'
```

salrashid123 commented:
@danawillow could you rerun the test and see if it passes?  I'm not really doing anything different to bootstrap impersonated credentials than that existing test.

danawillow commented:
Yup, it passes. I should have some time to do a final review later today or early tomorrow.

salrashid123 commented:
done; updated description;

thanks!   :)

danawillow commented:
I had to make a small change to the upstream one and I'm a bit nervous about merge conflicts, so I'm going to close this. The code will be merged in via https://github.com/GoogleCloudPlatform/magic-modules/pull/3739.
